### ISSUE TITLE
exchangedata is broken for opened files

### ISSUE THREAD

**Issue by aleksandr-skobelev-epam** [osxfuse/osxfuse#474]
If to keep opened file descriptors on files that is going to be exchange, the `exchangedata` operation won't perform properly. Here the output from a `test_exchange` utility on local disk:

    $ ./test_exchange /private/tmp
    File 3 has path "/private/tmp/file_exchange_1.txt".
    File 3 (length = 26, read = 26) contains "abcdefghijklmnopqrstuvwxyz".
    File 4 has path "/private/tmp/file_exchange_2.txt".
    File 4 (length = 26, read = 26) contains "ABCDEFGHIJKLMNOPQRSTUVWXYZ".
    -- Exchanged "/private/tmp/file_exchange_1.txt" and "/private/tmp/file_exchange_2.txt"
    File 3 has path "/private/tmp/file_exchange_2.txt".
    File 3 (length = 26, read = 26) contains "abcdefghijklmnopqrstuvwxyz".
    File 4 has path "/private/tmp/file_exchange_1.txt".
    File 4 (length = 26, read = 26) contains "ABCDEFGHIJKLMNOPQRSTUVWXYZ".
    -- Reopen "/private/tmp/file_exchange_1.txt" and "/private/tmp/file_exchange_2.txt"
    File 3 has path "/private/tmp/file_exchange_1.txt".
    File 3 (length = 26, read = 26) contains "ABCDEFGHIJKLMNOPQRSTUVWXYZ".
    File 4 has path "/private/tmp/file_exchange_2.txt".
    File 4 (length = 26, read = 26) contains "abcdefghijklmnopqrstuvwxyz".

After `exchangedata` call the file descriptors refer the old data and swap their paths.

In contrary here is output from the same utility on the mounted LoopbackFS volume:

    $ ./test_exchange  /Volumes/loop/
    File 3 has path "/Volumes/loop/file_exchange_1.txt".
    File 3 (length = 26, read = 26) contains "abcdefghijklmnopqrstuvwxyz".
    File 4 has path "/Volumes/loop/file_exchange_2.txt".
    File 4 (length = 26, read = 26) contains "ABCDEFGHIJKLMNOPQRSTUVWXYZ".
    -- Exchanged "/Volumes/loop/file_exchange_1.txt" and "/Volumes/loop/file_exchange_2.txt"
    File 3 has path "/Volumes/loop/file_exchange_1.txt".
    File 3 (length = 26, read = 26) contains "abcdefghijklmnopqrstuvwxyz".
    File 4 has path "/Volumes/loop/file_exchange_2.txt".
    File 4 (length = 26, read = 26) contains "ABCDEFGHIJKLMNOPQRSTUVWXYZ".
    -- Reopen "/Volumes/loop/file_exchange_1.txt" and "/Volumes/loop/file_exchange_2.txt"
    File 3 has path "/Volumes/loop/file_exchange_1.txt".
    File 3 (length = 26, read = 26) contains "abcdefghijklmnopqrstuvwxyz".
    File 4 has path "/Volumes/loop/file_exchange_2.txt".
    File 4 (length = 26, read = 26) contains "ABCDEFGHIJKLMNOPQRSTUVWXYZ".

After `exchangedata()` call the file descriptors refer to the old data and keep the old paths. Reopening shows that exchange has not been performed. 

The source of `test_exchange` utility is attached:
[test_exchange.c.zip](https://github.com/osxfuse/osxfuse/files/1706052/test_exchange.c.zip)

**Comment by Ajaybh** [comment]
Observing a crash in this functionality, not sure if the cause has to do anything with this defect.

### unique: 84, success, outsize: 32 unique: 82, opcode: EXCHANGE (63), nodeid: 648, insize: 84, pid: 54252 
### fuse internal error: node 32768 not found

Here's a snippet of some of the events that I found interesting:
_unique: 81, opcode: FSYNC (20), nodeid: 648, insize: 56, pid: 54252 fsync[105827995175168] datasync: 1 unique: 81, success, outsize: 16 unique: 81, opcode: OPEN (14), nodeid: 648, insize: 48, pid: 54252 open flags: 0x0 /Personal Folders/test01.doc.sb-5ad68c27-CBOFEK/~WRD0802 open[106102873056384] flags: 0x0 /Personal Folders/test01.doc.sb-5ad68c27-CBOFEK/~WRD0802 unique: 81, success, outsize: 32 unique: 84, opcode: OPEN (14), nodeid: 648, insize: 48, pid: 54252 open flags: 0x2 /Personal Folders/test01.doc unique: 84, success, outsize: 32 unique: 82, opcode: EXCHANGE (63), nodeid: 648, insize: 84, pid: 54252 fuse internal error: node 32768 not found_

Here's the complete log:
[Crash_DebugLogs.zip](https://github.com/osxfuse/osxfuse/files/1728423/Crash_DebugLogs.zip)

Also updated a slightly older issue (https://github.com/osxfuse/osxfuse/issues/356) with the same information. 

Hope this information is of help to fix the crash.

**Comment by bfleischer** [comment]
@aleksandr-skobelev-epam Thanks for the bug report and the super helpful test utility. The issue should be fixed by commit https://github.com/osxfuse/kext/commit/d5e4099ba108bf1f8813883b9251f8dfc948cdd4.

@Ajaybh I don't think your issue is related to Aleksandr's issue. Which file system are you using? Please try to reproduce the issue with the loopback file system, in case you are not already using loopback.

**Comment by aleksandr-skobelev-epam** [comment]
@bfleischer Thank you!

**Comment by Ajaybh** [comment]
Thanks @bfleischer for looking at the logs that I shared. The problem I reported is quite hard to reproduce. It happens with Word document editing and requires a long wait & a lot of save-as operations. Not sure if I would be able to easily reproduce the issue with loopback file system.
The issue has disappeared after we disabled the Exchange data call. 
I will try enabling the flow with your fix.

Thanks again!
